jujugui/static/gui/src/app/components/user-profile/budget-list.js

```diff
@@ -0,0 +1,184 @@
+/*
+This file is part of the Juju GUI, which lets users view and manage Juju
+environments within a graphical interface (https://launchpad.net/juju-gui).
+Copyright (C) 2016 Canonical Ltd.
+
+This program is free software: you can redistribute it and/or modify it under
+the terms of the GNU Affero General Public License version 3, as published by
+the Free Software Foundation.
+
+This program is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Affero
+General Public License for more details.
+
+You should have received a copy of the GNU Affero General Public License along
+with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+'use strict';
+
+YUI.add('budget-list', function() {
+
+  juju.components.BudgetList = React.createClass({
+    propTypes: {
+      listBudgets: React.PropTypes.func.isRequired,
+      user: React.PropTypes.object,
+    },
+
+    getInitialState: function() {
+      this.xhrs = [];
+
+      return {
+        budgetList: [],
+        loadingBudgets: false,
+      };
+    },
+
+    componentWillMount: function() {
+      this._getBudgets();
+    },
+
+    componentWillUnmount: function() {
+      this.xhrs.forEach((xhr) => {
+        xhr && xhr.abort && xhr.abort();
+      });
+    },
+
+    componentWillReceiveProps: function(nextProps) {
+      // If the user has changed then update the data.
+      var props = this.props;
+      var currentUser = props.user && props.user.user;
+      var nextUser = nextProps.user && nextProps.user.user;
+      if (nextUser !== currentUser) {
+        this._getBudgets();
+      }
+    },
+
+    /**
+      Get the budgets for the authenticated user.
+
+      @method _getBudgets
+    */
+    _getBudgets: function() {
+      // Delay the call until after the state change to prevent race
+      // conditions.
+      this.setState({loadingBudgets: true}, () => {
+        var xhr = this.props.listBudgets(this._getBudgetsCallback);
+        this.xhrs.push(xhr);
+      });
+    },
+
+    /**
+      Callback for the plans API call to get budgets.
+
+      @method _getBudgetsCallback
+      @param {String} error The error from the request, or null.
+      @param {Object} data The data from the request.
+    */
+    _getBudgetsCallback: function(error, data) {
+      this.setState({loadingBudgets: false});
+      if (error) {
+        if (error.indexOf('not found') === -1) {
+          // A "profile not found" error is expected, and it means the user
+          // does not have a credit limit yet. Notify any other errors.
+          // TODO huwshimi: notify the user with the error.
+          console.error('cannot retrieve budgets:', error);
+        }
+        return;
+      }
+      this.setState({budgetList: data.budgets});
```

Comment:
There are two set states in this method which could create a race condition. You could combine then (with a default if there is no budget data) or do the error check and second setState in a callback on the first setState.